### PR TITLE
Master l10n nl edi las

### PR DESCRIPTION
**Pull request by bfr-o**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account_edi_ubl/models/account_edi_format.py [hunk 1]
```diff
@@ -16,6 +15,17 @@
 class AccountEdiFormat(models.Model):
     _inherit = 'account.edi.format'
 
+    ####################################################
+    # Helpers
+    ####################################################
+
+    def _is_ubl(self, filename, tree):
+        return tree.tag == '{urn:oasis:names:specification:ubl:schema:xsd:Invoice-2}Invoice'
```
Comment (jco-odoo) [inline addons/account_edi_ubl/models/account_edi_format.py:23]: This will conflict with PE

File: addons/account_edi_ubl/models/account_edi_format.py [hunk 2]
```diff
@@ -24,19 +34,7 @@ def _import_ubl(self, tree, invoice):
...
-        def _get_ubl_namespaces():
-            ''' If the namespace is declared with xmlns='...', the namespaces map contains the 'None' key that causes an
-            TypeError: empty namespace prefix is not supported in XPath
-            Then, we need to remap arbitrarily this key.
-
-            :param tree: An instance of etree.
-            :return: The namespaces map without 'None' key.
-            '''
-            namespaces = tree.nsmap
-            namespaces['inv'] = namespaces.pop(None)
```
Comment (jco-odoo) [inline addons/account_edi_ubl/models/account_edi_format.py:36]: Are you sure this is not specific to UBL?

File: addons/account_edi_ubl/models/account_edi_format.py [hunk 3]
```diff
@@ -159,3 +157,85 @@ def _find_value(xpath, element=tree):
...
+
+    def _get_ubl_values(self, invoice):
+        ''' Get the necessary values to generate the XML. These values will be used in the qweb template when
+        rendering. Needed values differ depending on the implementation of the UBL, as (sub)template can be overriden
+        or called dynamically.
+        :returns:   a dictionary with the value used in the template has key and the value as value.
+        '''
+        def format_monetary(amount):
+            # Format the monetary values to avoid trailing decimals (e.g. 90.85000000000001).
+            return float_repr(amount, invoice.currency_id.decimal_places)
```
Comment (jco-odoo) [inline addons/account_edi_ubl/models/account_edi_format.py:173]: I think we need the currency.round as well in a lot of cases e.g. with the calculation of discounts.

File: addons/account_edi_ubl/models/account_edi_format.py [hunk 4]
```diff
@@ -159,3 +157,85 @@ def _find_value(xpath, element=tree):
...
+
+        return {
+            **invoice._prepare_edi_vals_to_export(),
+            'ubl_version': 2.1,
+            'type_code': 380 if invoice.move_type == 'out_invoice' else 381,
+            'payment_means_code': 42 if invoice.journal_id.bank_account_id else 31,
+            'bank_account': invoice.partner_bank_id,
+            'format_monetary': format_monetary,
+            'customer_vals': {'partner': invoice.commercial_partner_id},
+            'supplier_vals': {'partner': invoice.company_id.partner_id.commercial_partner_id},
```
Comment (jco-odoo) [inline addons/account_edi_ubl/models/account_edi_format.py:183]: This is confusing in so many ways.  supplier_vals should be company_vals or something.  (or emittor_vals e.g.)

File: addons/account_edi_ubl_bis3/models/account_edi_format.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    'PL': 9945,
+    'PT': 9946,
+    'RO': 9947,
+    'RS': 9948,
+    'SI': 9949,
+    'SK': 9950,
+    'SM': 9951,
+    'TR': 9952,
+    'VA': 9953,
+
```
Comment (jco-odoo) [inline addons/account_edi_ubl_bis3/models/account_edi_format.py:44]: why the empty lines?

File: addons/account_edi_ubl_bis3/models/account_edi_format.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    def _get_bis3_values(self, invoice):
+        values = super()._get_ubl_values(invoice)
+        values.update({
+            'customization_id': 'urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0',
+            'profile_id': 'urn:fdc:peppol.eu:2017:poacc:billing:01:1.0',
+        })
+
+        all_tax_detail_per_tax = {}
+        for line_vals in values['invoice_line_vals_list']:
+            line_vals['price_subtotal_with_no_tax_closing'] = line_vals['line'].price_subtotal
```
Comment (jco-odoo) [inline addons/account_edi_ubl_bis3/models/account_edi_format.py:70]: Why not have it right from the line_vals then?  In generic it is used a certain way and in bis3 differently?

File: addons/l10n_nl/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# coding: utf-8
+from odoo import fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    l10n_nl_kvk = fields.Char(string='KVK-nummer')
```
Comment (jco-odoo) [inline addons/l10n_nl/models/res_partner.py:8]: company_registry normally but we don't have that field on the partner (so, you are sure you need this for every customer and not just for the company?)

File: addons/l10n_nl/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+    l10n_nl_oin = fields.Char(string='Organisatie Indentificatie Nummer')
```
Comment (jco-odoo) [inline addons/l10n_nl/models/res_partner.py:9]: Only when you send it digitally?

File: addons/l10n_nl_edi/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+{
+    'name': 'Netherlands - E-Invoicing (NLCIUS 1.0.3)',
+    'version': '0.1',
+    'category': 'Accounting/Localizations/EDI',
+    'summary': 'E-Invoicing, Universal Business Language (NLCIUS 1.0.3), nlc',
+    'description': """
+NLCIUS is the Dutch implementation of EN 16931 norm (with some limitations). Both for UBL and UN / CEFACT XML CII.
```
Comment (jco-odoo) [inline addons/l10n_nl_edi/__manifest__.py:8]: I would not mind a little description about how the NLCIUS is the NL dialect of the BIS 3.0 and if it can be sent standalone or you need the PEPPOL network for it.

File: odoo/addons/base/models/res_partner.py [hunk 10]
```diff
@@ -198,6 +198,7 @@ def default_get(self, default_fields):
     city = fields.Char()
     state_id = fields.Many2one("res.country.state", string='State', ondelete='restrict', domain="[('country_id', '=?', country_id)]")
     country_id = fields.Many2one('res.country', string='Country', ondelete='restrict')
+    country_code = fields.Char(related='country_id.code', string="Country Code")
```
Comment (jco-odoo) [inline odoo/addons/base/models/res_partner.py:201]: This should go into a separate commit.

**Review by jco-odoo — CHANGES_REQUESTED** [review]
@bfr-o After a first pass through the code, I really do not see the benefit of the parent_id on the format.  It is just confusing me.

File: addons/account_edi_ubl_bis3/models/account_edi_format.py (hunk 6 above)
Comment (bfr-o) [reply addons/account_edi_ubl_bis3/models/account_edi_format.py:70]: In generic it's not used, this is something specific to BIS3

File: addons/l10n_nl/models/res_partner.py (hunk 7 above)
Comment (bfr-o) [reply addons/l10n_nl/models/res_partner.py:8]: What is company_registry ? This seems like a field where you can put anything, with no real definition.

In nl, there can be multiple "specific numbers" for the same company : oin or kvk. For the seller, kvk is mandatory, but for customer either kvk or oin is fine but you need one.

File: addons/l10n_nl/models/res_partner.py (hunk 8 above)
Comment (bfr-o) [reply addons/l10n_nl/models/res_partner.py:9]: Can you elaborate ? :D Otherwise, see my other comment https://github.com/odoo/odoo/pull/68514#discussion_r604769021

File: addons/account_edi_ubl_bis3/models/account_edi_format.py (hunk 5 above)
Comment (bfr-o) [reply addons/account_edi_ubl_bis3/models/account_edi_format.py:44]: For formating. The codes are incremental, but there are holes. This is easier to read, and will be easier to add codes when we will support them or the holes are filled by the standard.

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 4 above)
Comment (bfr-o) [reply addons/account_edi_ubl/models/account_edi_format.py:183]: I think company/partner is more confusing than customer/supplier. If we want to support bills and invoices, the use of company/partner is so confusing when you need to match them with odoo's fields. Customer/supplier is the name defined by BIS3 (see the xml), so I think this is as clear as can be.

File: addons/account_edi/models/account_edi_format.py [hunk 11]
```diff
@@ -406,15 +404,45 @@ def _update_invoice_from_attachment(self, attachment, invoice):
...
+            edi_format = edi_format.parent_id
+        return edi_code in codes
+
     ####################################################
     # Import helpers
     ####################################################
+    def _get_namespaces(self, tree):
+        ''' If the namespace is declared with xmlns='...', the namespaces map contains the 'None' key that causes an
+        TypeError: empty namespace prefix is not supported in XPath
+        Then, we need to remap arbitrarily this key.
```
[the remark this answers is not in the document]
Comment (bfr-o) [reply addons/account_edi/models/account_edi_format.py:432]: I agree this is fuzzy... This is actually the case for every xml where there is a namespace with empty prefix (xmlns=something), which is often the case (not only ubl). The namespace manipulation in etree is a pain, for `find´, `find_all` and `xpath` you need to give the namespaces. lxml eased things a bit by providing an easy way to retrieve the namespaces by using `root.nsmap`, however this returns also the namespaces with empty prefix, which are not supported by these methods (generates an error). Hence the need to remap it.

You should do that every time you want to use theses methods, which in the EDI happens very often. Previously, this method was defined as an inner method of a lot of import methods. Rather than defining multiple times, I think it's good to have this helper that is defined once and can be used everywhere. In the future I would like to add some more helpers for the more common stuff that we need to do in the edi (finding values mostly). 

Again, I agree, this is a pain that we have to do that this way. If you have a better suggestion, I'm open.

[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/account_edi/models/account_edi_format.py:432]: The reasons seem legit, but the docstring is difficult to understand.  So we filter empty from root.nsmap?  (would not lxml have a helper method itself?)

[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/account_edi/models/account_edi_format.py:432]: The explanation you gave here is a lot clearer.

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 2 above)
Comment (bfr-o) [reply addons/account_edi_ubl/models/account_edi_format.py:36]: See https://github.com/odoo/odoo/pull/68514#discussion_r604777776

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 1 above)
Comment (smetl) [reply addons/account_edi_ubl/models/account_edi_format.py:23]: No import in PE so no conflict

Comment (bfr-o) [reply addons/account_edi_ubl/models/account_edi_format.py:23]: We don't import invoices on PE.

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 3 above)
Comment (smetl) [reply addons/account_edi_ubl/models/account_edi_format.py:173]: In my mind, every monetary contained in the values are already rounded

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 4 above)
Comment (jco-odoo) [reply addons/account_edi_ubl/models/account_edi_format.py:183]: Ok, I did not see they used that term in the XML

File: addons/l10n_nl/models/res_partner.py (hunk 8 above)
Comment (jco-odoo) [reply addons/l10n_nl/models/res_partner.py:9]: Checking online, this field is only used for sending digitally.  (You need this number if you want to send it electronically to the government and I don't think we support this yet.)

File: addons/l10n_nl/models/res_partner.py (hunk 7 above)
Comment (jco-odoo) [reply addons/l10n_nl/models/res_partner.py:8]: KVK is like ondernemingsnummer in Belgium and is normally put in that field. But if we need it for the customer, we need this field.

File: addons/account_edi_ubl/models/account_edi_format.py (hunk 3 above)
Comment (jco-odoo) [reply addons/account_edi_ubl/models/account_edi_format.py:173]: Yes, but I started doubting with all the tax_detail_vals we send, but it seems to be rounded correctly there.  (at least what is in the xml)

File: addons/account_edi/models/account_edi_format.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (smetl) [reply addons/account_edi/models/account_edi_format.py:432]: Use `find({*}<my_tag>)`

[the remark this answers is not in the document]
Comment (bfr-o) [reply addons/account_edi/models/account_edi_format.py]: Instead we skip the namespaces using `element.find('./{*}myTag')`

**Comment by smetl** [comment]
@robodoo r+ rebase-ff
